ocaml/xapi/xapi_host.ml

```diff
@@ -1425,15 +1472,30 @@ let reset_server_certificate ~__context ~host =
         xs
   in
   let cn = ip in
-  Gencertlib.Selfcert.host cn alt_names xapi_ssl_pem ip ;
+  let certificate, root_fingerprint =
+    Gencertlib.Selfcert.generate cn alt_names xapi_ssl_pem ip
+  in
+
+  let cert = X509.Certificate.encode_pem certificate |> Cstruct.to_string in
+
+  (* install self-signed certificate to the pool's trusted CA certs.
+     Uses a uuid as name to avoid duplicate names. This creates a
+     certificate records to the database as a pool certificate.
+     Since we just generated the private key randomly we know the fingerptint
+     in the database will be unique and we can replace it. *)
+  let name = Uuid.(to_string (make_uuid ())) in
+
+  Helpers.call_api_functions ~__context (fun rpc session_id ->
+      Client.Client.Pool.install_ca_certificate rpc session_id name cert) ;
+
+  let self = Db.Certificate.get_by_uuid ~__context ~uuid:name in
+  Db.Certificate.set_type ~__context ~self ~value:`host_and_pool ;
+  Db.Certificate.set_host ~__context ~self ~value:host ;
+  Db.Certificate.set_validated_by ~__context ~self ~value:[self] ;
+
   (* Reset stunnel to try to restablish TLS connections *)
   Xapi_mgmt_iface.reconfigure_stunnel ~__context ;
-  (* Delete records of the server certificate in this host *)
-  let expr =
-    Db_filter_types.(Eq (Field "host", Literal (Ref.string_of self)))
-  in
-  Db.Certificate.get_refs_where ~__context ~expr
-  |> List.iter (fun self -> Db.Certificate.destroy ~__context ~self)
+  uninstall_outdated_certificates ~__context current_server_certs
```

Comment:
If we restart stunnel before uninstalling the old certificates, then won't the pool potentially have some connections which are verified using the old certificates? (this might be bad for long running connections established at start up, but equally might not be a concern for us, since those 'old' certificates are still valid)

(In reality this is probably not an issue, because I believe the call to `reconfigure_stunnel` is threaded, and so by the time stunnel is back up the old certificates will have been removed)

Reply:
> won't the pool potentially have some connections which are verified using the old certificates?

I don't think so: any new connection established after the stunnel reset will offer the new certificate to clients. This is why it's safe to delete the server certificates from the pool's trust anchors at this point and not safe to do before.

> I believe the call to reconfigure_stunnel is threaded, and so by the time stunnel is back up the old certificates will have been removed

That might be a problem. in this case the call must be made blocking.